modules/macros/pages/link-macro-attribute-parsing.adoc

```diff
@@ -12,8 +12,8 @@ Normally, the whole text between the square brackets of a link macro is treated
 https://discuss.asciidoctor.org[Discuss Asciidoctor]
 ----
 
-However, if the text contains both an equals sign (`=`) and a comma (`,`), and the comma comes after the equals sign, the text is parsed as an attribute list.
-In this case, the linked text is the first positional attribute.
+However, if the text contains an equals sign (`=`), the text is parsed as an xref:attributes:element-attributes.adoc#attribute-list[attribute list].
+If the first attribute is enclosed in double quotes (`"`) or contains no equals sign, it forms the linked text.
```

Comment:
This is suggesting that the linked text has to be enclosed in double quotes, which is not true.

Rewrite as:

"In this case, the linked text is the first positional (unnamed) attribute. The linked text must be enclosed in double quotes if it contains either a comma or an equal sign that occurs before the first space."

Reply:
The rule for when the linked text must be enclosed in double quotes is as follows:

1. An equal sign is present anywhere in the attrlist (the area between the square brackets) AND
2. a) The linked text contains a comma OR b) the linked text contains an equal sign that occurs before the first space

The really weird edge case is `foo = bar,baz`. Here the equal sign is after the first space, but that's because the space around the equal sign gets trimmed. So it's effectively `foo=bar,baz`, which matches 1b.

Reply:
Btw, I realize these exceptions are kind of nuts. This was a compromise to ease migration since AsciiDoc Python did not support attributes in a link macro. When we standardize it, we may just decide to always parse as an attribute list, which would then be consistent with the rules everywhere for the interplay between positional and named attributes.

Reply:
I'm not sure you have this right.  In the IntelliJ preview, `link:https:example.com/foo.html[This = that]` does not show as a positional attribute.

Reply:
There's no way for a user to know about trimming spaces.... we have to provide some guidance based on what they supply, not after some amount of parsing.

Reply:
In any case, the code clearly looks for an '=' anywhere and if found parses as an attribute list.

Reply:
AFAICT with attribute list parsing if the first attribute is unquoted and has an equal sign anywhere it is not positional.

Reply:
Right, that's what I'm saying when I stated the exception. If you consider the equals sign trimmed, if there is still a space before that, then it isn't parsed as a named attribute. So `foo = bar` is a named attribute, whereas `foo bar = baz` is not.

Reply:
So perhaps we can say "trimmed equal sign" in the rule and then it will be accurate. It's a bit strange to say it, but technically correct.

Reply:
"In this case, the linked text is the first positional (unnamed) attribute. The linked text must be enclosed in double quotes if it contains either a comma or a trimmed equal sign that occurs before the first space."

Reply:
> AFAICT with attribute list parsing if the first attribute is unquoted and has an equal sign anywhere it is not positional.

See my exception. It's if it has a trimmed equal sign that occurs before the first space. This is an accurate statement.

Reply:
We can't expect users to know to trim space around equals signs.

Reply:
Maybe I'm confused.... let me think through this again.

Reply:
OK, your `foo = bar` does not contain an equals sign that occurs before the first space, and has no comma, so it doesn't need to be quoted by your rules. However, it does need to be quoted. Did I miss something? There are a lot of negatives here, they are hard to count mod 2.

Reply:
You are not getting what I'm saying. When the user looks at the equal sign, and disregards the spaces immediately around it, then the rule is applied. I understand this is confusing, but we cannot change the rules because it is hard to understand. This is the spec and needs to be written as such. So the rule is as I stated it. If you have a better way of writing "trimmed equals sign" then I'm open to that.

Reply:
`foo = bar` is equivalent to `foo=bar`. The spaces around the equal sign are insignificant.

Reply:
yikes.... `1=>2 posits the problem of inequality` is a positional attribute.

Reply:
I actually think we should describe the rules on the definition of 'attribute list' and refer to that from here.

Reply:
Here I think we should say something like, "a simplified rule is to enclose the linked text in quotes when it contains a comma or equals sign".

Reply:
Ruby asciidoctor parses my example correctly, asciidoctor.js doesn't.  See https://github.com/asciidoctor/asciidoctor.js/issues/1178.

We need to explain how attribute list parsing works where we explain attribute lists, not here where we're just using them.  I'm going to rewrite this page from that point of view.  Then we can work on the attribute list page.

Reply:
That's really strange that they are different. It's possible it's a regex compatibility issue. Otherwise, the code is exactly the same.

Reply:
I realized there is another, perhaps better, way to explain when the double quotes are needed.

> If the linked text contains an equal sign, and the text that precedes that equals sign (ignoring spaces immediately adjacent to the equals sign) is a valid attribute name (letters and numbers only), then it will be parsed as an attribute and the linked text will be empty.

Reply:
And yes, we can say "when in doubt, enclose the linked text in double quotes". But it becomes pretty obvious when you don't get the text that you expect that the double quotes could be relevant.